### PR TITLE
feat: adding ability to download models

### LINKED ISSUE #10: Integrate HuggingfaceModelDownloader
We can use HuggingfaceModelDownloader to allow for direct download of models from the huggingface hub.
It would be good to include a build tag so that one can still build the project without this facility, but having it would make it easier to directly load an onnx model.

### PR DESCRIPTION
**Pull request by riccardopinosio**
Adds the ability to download models directly from huggingface.

Changes:
session object now has .DownloadModel method to download a model from huggingface
cli can now download model directly with the --model parameter. Chain is: first check if --model is an existing path, if not check if --model is a model name that exists at $HOME/hugot/models, if not try to download --model to $HOME/hugot/models and use it.

Note: model download depends on huggingfaceModelDownloader. This functionality can be disabled at build time with the -t NODOWNLOAD tag.

closes #10

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -84,6 +86,14 @@ var runCommand = &cli.Command{
...
 
 		var onnxLibraryPathOpt hugot.SessionOption
+		userDir, err := os.UserHomeDir()
+		hugoDir := util.PathJoinSafe(userDir, "hugot")
+
+		if err != nil {
+			return err
+		}
+
+		fmt.Println(modelPath)
```
Comment (RJKeevil) [inline cmd/main.go:96]: Remove print statement

File: cmd/main.go [hunk 2]
```diff
@@ -84,6 +86,14 @@ var runCommand = &cli.Command{
 	Action: func(ctx *cli.Context) error {
 
 		var onnxLibraryPathOpt hugot.SessionOption
+		userDir, err := os.UserHomeDir()
```
Comment (RJKeevil) [inline cmd/main.go:89]: Perhaps allow the user to provide a dir, falling back to home if blank?

File: downloader.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+		if err := hfd.DownloadModel(modelName, false, options.SkipSha, false, destination, options.Branch, options.ConcurrentConnections, options.AuthToken, !options.Verbose); err != nil {
+			fmt.Printf("Warning: attempt %d / %d failed, error: %s\n", i+1, options.MaxRetries, err)
+			time.Sleep(time.Duration(options.RetryInterval) * time.Second)
+			continue
+		}
+		fmt.Printf("\nDownload of %s completed successfully\n", modelName)
+		return modelPath, nil
+	}
+	return "", fmt.Errorf("failed to download %s after %d attempts", modelName, options.MaxRetries)
+}
```
Comment (RJKeevil) [inline downloader.go:51]: Might be good to check downloaded model actualy has the required files, e.g. tokenizer and onnx?

File: cmd/main.go [hunk 4]
```diff
@@ -110,6 +120,37 @@ var runCommand = &cli.Command{
...
+		// find or download the model
+		ok, err := util.FileSystem.Exists(ctx.Context, modelPath)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			modelsDir := util.PathJoinSafe(hugoDir, "models")
+			modelName := modelPath
+			if strings.Contains(modelName, ":") {
+				modelName = strings.Split(modelPath, ":")[0]
```
Comment (RJKeevil) [inline cmd/main.go:132]: slightly better if you use modelName here, in case we chain more modifications above

Comment (riccardopinosio) [reply cmd/main.go:132]: I'm actually going to just error if the --model is not a full model path and contains ":". This ":" thing is used for model/dataset filters, not super important to support it at the moment.

File: downloader.go (hunk 3 above)
Comment (riccardopinosio) [reply downloader.go:51]: done. It's a bit of a question whether we want to be strict, i.e. there must be a "model.onnx" file, or whether we try to load the first ".onnx" file we find

Comment (RJKeevil) [reply downloader.go:60]: Yeh difficult one, there is no real standard here, sometimes you get model and model quantized. Probs load if exactly one exists, else they must provide the file name as a param?

**Pull request merged**